File: Src/TeaDriven.Graphology.Tests/Visualization/DefaultGetNextLetterCodeTests.cs
using Ploeh.AutoFixture.Idioms;
using TeaDriven.Graphology.Tests.Conventions;
using TeaDriven.Graphology.Visualization;
using Xunit;
using Xunit.Extensions;

namespace TeaDriven.Graphology.Tests.Visualization
{
    public class DefaultGetNextLetterCodeTests
    {
        [Theory, AutoNSubstituteData]
        public void Constructor_HasNullGuards(GuardClauseAssertion assertion)
        {
            assertion.Verify(typeof(DefaultGetNextLetterCode).GetConstructors());
        }

        [Theory, AutoNSubstituteData]
        public void Constructor_ReturnsIGetNextLetterCode(DefaultGetNextLetterCode sut)
        {
            Assert.IsAssignableFrom<IGetNextLetterCode>(sut);
        }

        [Theory]
        [InlineAutoNSubstituteData("A", "B")]
        [InlineAutoNSubstituteData("M", "N")]
        public void From_ReturnsCorrectNextLetterCode(string previous, string expectedResult, DefaultGetNextLetterCode sut)
        {
            // Arrange

            // Act
            var result = sut.From(previous);

            // Assert
            Assert.Equal(expectedResult, result);
        }
    }
}

File: Src/TeaDriven.Graphology.TestsF/DefaultGetNextLetterCodeTests.fs
namespace TeaDriven.Graphology.Tests

open Xunit
open Xunit.Extensions
open Ploeh.AutoFixture.Idioms
open FsUnit.Xunit

module DefaultGetNextLetterCodeTests =
    open TeaDriven.Graphology.Visualization

    [<Theory; AutoFoqData>]
    let ``Constructors should have null guards`` (assertion : GuardClauseAssertion) =
        assertion.Verify (typeof<DefaultGetNextLetterCode>.GetConstructors())

    [<Theory; AutoFoqData>]
    let ``Class should implement IGetNextLetterCode`` (sut : DefaultGetNextLetterCode) =
        Assert.IsAssignableFrom<IGetNextLetterCode> sut

    [<Theory; AutoFoqData>]
    let ``From() should have null guards`` (assertion: GuardClauseAssertion) =
        assertion.Verify (typeof<DefaultGetNextLetterCode>.GetMethod "From")

    [<Theory>]
    [<InlineAutoFoqData("A", "B")>]
    [<InlineAutoFoqData("M", "N")>]
    [<InlineAutoFoqData("Z", "AA")>]
    [<InlineAutoFoqData("AA", "AB")>]
    [<InlineAutoFoqData("AZ", "BA")>]
    [<InlineAutoFoqData("ZZ", "AAA")>]
    [<InlineAutoFoqData("AAAZ", "AABA")>]
    let ``For() returns correct next letter code`` (oldCode : string, expectedResult : string, sut : DefaultGetNextLetterCode) =
        let result = sut.From oldCode

        result |> should equal expectedResult